File: src/main/java/dev/robothanzo/werewolf/model/PoliceSession.java
package dev.robothanzo.werewolf.model;

import dev.robothanzo.werewolf.database.documents.Session;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import net.dv8tion.jda.api.entities.Message;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class PoliceSession {
    private long guildId;
    private long channelId;
    private Session session;
    @Builder.Default
    private State state = State.NONE;
    @Builder.Default
    private long stageEndTime = 0;
    @Builder.Default
    private Map<Integer, Candidate> candidates = new ConcurrentHashMap<>();
    @Builder.Default
    private Message message = null;

    public enum State {
        NONE, ENROLLMENT, SPEECH, UNENROLLMENT, VOTING, FINISHED;

        public boolean canEnroll() {
            return this == ENROLLMENT;
        }

        public boolean canQuit() {
            return this == ENROLLMENT || this == UNENROLLMENT;
        }
    }
}

File: src/main/kotlin/dev/robothanzo/werewolf/model/PoliceSession.kt
package dev.robothanzo.werewolf.model

import dev.robothanzo.werewolf.database.documents.Player
import dev.robothanzo.werewolf.database.documents.Session
import net.dv8tion.jda.api.entities.Message

class PoliceSession(
    guildId: Long,
    channelId: Long,
    session: Session,
    var state: State = State.NONE,
    var stageEndTime: Long = 0,
    message: Message? = null,
    finishedCallback: (() -> Unit)? = null
) : AbstractPoll(guildId, channelId, session, message, finishedCallback) {
    enum class State {
        NONE, ENROLLMENT, SPEECH, UNENROLLMENT, VOTING, FINISHED;

        fun canEnroll(): Boolean {
            return this == ENROLLMENT
        }

        fun canQuit(): Boolean {
            return this == ENROLLMENT || this == UNENROLLMENT
        }
    }

    override fun isEligibleVoter(player: Player): Boolean {
        // Only players who never enrolled (i.e., not present in candidates map) are eligible
        return player.alive && !candidates.containsKey(player.id)
    }
}
